### DISCUSSION TITLE
Search options in the filter.

### DISCUSSION THREAD

**Discussion by philipbaginski** [laravel/nova-issues#4523] [General]
Hi!

This issue/request is labeled as implemented, but I can not find it in docs: https://github.com/laravel/nova-issues/issues/1894

Is there any way to have search options in the filter with many options based on belongsTo?

**Comment by crynobone** [comment]
In Nova 4 just write

```php
BelongsTo::make('User')->filterable()->searchable(),
```

**Reply by philipbaginski** [reply]
> In Nova 4 just write
> 
> ```
> BelongsTo::make('User')->filterable()->searchable(),
> ```

Can I use BelongsTo in the filters?

I need to use here:

    /**
     * Get the filter's available options.
     *
     * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
     * @return array
     */
    public function options(NovaRequest $request)
    {
        return User::orderBy('name', 'asc')->pluck('id', 'name');
    }

**Reply by juse-less** [reply]
Nova 4 comes with many pre-built filters for various `Field`s, like `Text`, `Boolean`, `BelongsTo`, etc ([documentation](https://nova.laravel.com/docs/4.0/resources/fields.html#filterable-fields)).

So, calling `BelongsTo::make('User')->filterable()` will show a filter alongside all other filters in the filter-popup.
Also applying `searchable()` on that makes the filter searchable (if the filter supports it).
So, if you copy @crynobone's example and paste as your `BelongsTo` field (in the `Resource`/`Lens` `fields()` method), it'll work out of the box.
If you had any custom filter to filter on `BelongsTo` resources, that one can be removed from the `filters()` method in the `Resource`/`Lens` classes.

I.e.
```php
class MyResource extends NovaResource
{
    public function fields(): array
    {
        return [
            BelongsTo::make('User')->filterable()->searchable(),
        ];
    }
}
```

If you have more custom needs, then you may need to create a custom filter to account for that logic, though.

**Reply by philipbaginski** [reply]
> Nova 4 comes with many pre-built filters for various `Field`s, like `Text`, `Boolean`, `BelongsTo`, etc ([documentation](https://nova.laravel.com/docs/4.0/resources/fields.html#filterable-fields)).
> 
> So, calling `BelongsTo::make('User')->filterable()` will show a filter alongside all other filters in the filter-popup. Also applying `searchable()` on that makes the filter searchable (if the filter supports it). So, if you copy @crynobone's example and paste as your `BelongsTo` field (in the `Resource`/`Lens` `fields()` method), it'll work out of the box. If you had any custom filter to filter on `BelongsTo` resources, that one can be removed from the `filters()` method in the `Resource`/`Lens` classes.
> 
> I.e.
> 
> ```
> class MyResource extends NovaResource
> {
>     public function fields(): array
>     {
>         return [
>             BelongsTo::make('User')->filterable()->searchable(),
>         ];
>     }
> }
> ```
> 
> If you have more custom needs, then you may need to create a custom filter to account for that logic, though.

Yes, but filterable makes filter I can not detach from filter popup. Than I need to create filter, not use filterable method. And if I have a filter file and options:

/**
 * Get the filter's available options.
 *
 * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
 * @return array
 */
public function options(NovaRequest $request)
{
    return User::orderBy('name', 'asc')->pluck('id', 'name');
}

There is no method to make it searchable.

**Reply by juse-less** [reply]
Ahh. I see. That makes it a bit trickier.
The `BelongsTo` `Field` is just that - a `Field`, and not a `Filter` in itself.
But maybe you could look up how the `Filter` used in `BelongsTo` is being used/constructed, and maybe try replicate the behaviour?

I haven't tried that before, so it's unfortunately unknown territory for me right now. 🙁
